File: ObjC/CustomTemplatesExample/CustomTemplatesExample/SliderTemplate/SliderMessageTemplate.m
//
//  SliderMessageTemplate.m
//  CustomTemplatesExample
//
//  Created by Nikola Zagorchev on 27.09.20.
//  Copyright © 2022 Nikola Zagorchev. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "SliderMessageTemplate.h"
#import "SliderViewController.h"
#import <Leanplum/LPActionContext.h>

#pragma mark - Arguments
static NSString *const Name = @"Slider";
static NSString *const TitleColorArg = @"Title Color";
static NSString *const TitlesArg = @"Titles";
static NSString *const ImagesArg = @"Images";

static NSString *const StoryboardName = @"SliderMessageTemplateStoryboard";
static NSString *const SliderViewControllerName = @"SliderViewController";

@implementation SliderMessageTemplate

#pragma mark - Definition
+ (void)defineAction {
    NSArray *titles = @[@"Personalize every message", @"Mobile App & Website Inbox Messages"];
    NSString *keyFormat = @"Title %d";
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    for (int i = 0; i < titles.count; i++) {
        NSString *key = [NSString stringWithFormat:keyFormat, i + 1];
        dict[key] = titles[i];
    }
    
    __block UIViewController *sliderViewController;
    BOOL (^presentHandler)(LPActionContext *) = ^(LPActionContext *context) {
        @try {
            UIViewController *topController = [self topViewController];
            sliderViewController = [SliderMessageTemplate viewControllerWithContext:context];
            [topController presentViewController:sliderViewController animated:YES completion:nil];
            
            return YES;
        }
        @catch (NSException *exception) {
            return NO;
        }
    };
    
    BOOL (^dismissHandler)(LPActionContext *) = ^(LPActionContext *context) {
        if (sliderViewController) {
            [sliderViewController dismissViewControllerAnimated:YES completion:^{
                [context actionDismissed];
            }];
            return YES;
        }
        return NO;
    };
    
    NSArray *args = @[
        [LPActionArg argNamed:TitleColorArg withColor:UIColor.blackColor],
        [LPActionArg argNamed:TitlesArg withDict:dict],
        [LPActionArg argNamed:ImagesArg withDict:@{}]
    ];
    
    [Leanplum defineAction:Name
                    ofKind:kLeanplumActionKindMessage
             withArguments:args
               withOptions:@{}
            presentHandler:presentHandler
            dismissHandler:dismissHandler];
}

#pragma mark - Presentation
+ (UIViewController *)viewControllerWithContext:(LPActionContext *)context {
    NSBundle *bundle = [NSBundle bundleForClass:[SliderViewController class]];
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:StoryboardName bundle:bundle];
    SliderViewController *vc = [storyboard instantiateViewControllerWithIdentifier:SliderViewControllerName];
    vc.modalPresentationStyle = UIModalPresentationFullScreen;
    
    NSDictionary *dictTitles = [context dictionaryNamed:TitlesArg];
    NSDictionary *dictImages = [context dictionaryNamed:ImagesArg];
    
    vc.slideTitleColor = [context colorNamed:TitleColorArg];
    vc.slideTitles = [self sortValuesByKeys:dictTitles];
    vc.slideImages = [self sortValuesByKeys:dictImages];
    
    return vc;
}

+ (UIViewController *)topViewController {
    UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
    while (topController.presentedViewController) {
        topController = topController.presentedViewController;
    }
    
    return topController;
}

+ (NSArray *)sortValuesByKeys:(NSDictionary *)dict {
    NSArray *sortedKeys = [[dict allKeys] sortedArrayUsingSelector: @selector(caseInsensitiveCompare:)];
    NSArray *objects = [dict objectsForKeys: sortedKeys notFoundMarker: [NSNull null]];
    
    return objects;
}

@end

File: Swift/CustomTemplatesExample/CustomTemplatesExample/Slider/SliderMessageTemplate.swift
//
//  SliderMessageTemplate.swift
//  CustomTemplatesExample
//
//  Created by Nikola Zagorchev on 6.10.20.
//  Copyright © 2022 Nikola Zagorchev. All rights reserved.
//

import Leanplum

class SliderMessageTemplate {
    
    // MARK: Arguments
    static let Name = "Slider"
    static let TitleColorArg = "Title Color"
    static let TitlesArg = "Titles"
    static let ImagesArg = "Images"
    
    static let StoryboardName = "SliderMessageTemplateStoryboard"
    static let SliderViewControllerName = "SliderViewController"
    
    // MARK: Definition
    static func defineAction() {
        var viewController: UIViewController?
        
        let present: LeanplumActionBlock = { (context: ActionContext) in
            viewController = self.viewControllerWithContext(with: context)
            
            guard let viewController = viewController,
                  let topViewController = topViewController else { return false }
            
            topViewController.present(viewController, animated: true, completion: nil)
            return true
        }
        
        let dismiss: LeanplumActionBlock = { (context: ActionContext) in
            if let viewController = viewController {
                viewController.dismiss(animated: true, completion: {
                    context.actionDismissed()
                })
                return true
            }
            
            return false
        }
        
        let titles = ["Personalize every message", "Mobile App & Website Inbox Messages"]
        
        var dict:Dictionary = Dictionary<String, Any>.init()
        for i in 0...titles.count - 1 {
            let key = "Title \(i + 1)"
            dict[key] = titles[i]
        }
        
        Leanplum.defineAction(name: SliderMessageTemplate.Name, kind: .message,
                              args: [ActionArg.init(name: SliderMessageTemplate.TitleColorArg, color: UIColor.blue),
                                     ActionArg.init(name: SliderMessageTemplate.TitlesArg, dictionary: dict),
                                     ActionArg.init(name: SliderMessageTemplate.ImagesArg, dictionary: [:])],
                              present: present,
                              dismiss: dismiss)
    }
    
    // MARK: Presentation
    static func viewControllerWithContext(with context: ActionContext) -> UIViewController {
        let bundle = Bundle.init(for: SliderViewController.self)
        let storyboard = UIStoryboard.init(name: StoryboardName, bundle: bundle)
        let vcs = storyboard.instantiateViewController(withIdentifier: SliderViewControllerName)
        let vc = vcs as! SliderViewController
        vc.modalPresentationStyle = .fullScreen
        
        let titlesDict = context.dictionary(name: TitlesArg) as! [String:Any]
        let imagesDict = context.dictionary(name: ImagesArg) as! [String:Any]
        
        let titles = titlesDict.keys.sorted().compactMap{ titlesDict[$0] } as! [String]
        let images = imagesDict.keys.sorted().compactMap{ imagesDict[$0] } as! [String]
        
        vc.slideTitles = titles
        vc.slideImages = images
        vc.slideTitleColor = context.color(name: TitleColorArg)
        
        return vc
    }
    
    class var topViewController: UIViewController? {
        var topController = UIApplication.shared.keyWindow!.rootViewController
        while (topController?.presentedViewController != nil) {
            topController = topController?.presentedViewController;
        }
        return topController
    }
}
